### DISCUSSION TITLE
Improve saem model

### DISCUSSION THREAD

**Discussion by afee36912** [nlmixr2/nlmixr2#247] [Q&A] [answered]
Hi Matt and all,

I have a couple questions about nlmixr2 and saem.

I am running a saem model on some data, but the model is simply not fitting well. For some parameters, I do not know a close-to-optimal tV so I am kind of guessing as to what would be an eta value to correspond. What method do you recommend to optimize parameter fitting? What should eta values be set to in the guessing stages of nlmixr modeling? I have tried setting nBurn and nEm to set the maximum # of iterations in saem fitting, but I do not have a good understanding of these parameters. What do these parameters do? What impact do they have? Is it beneficial to be pre-setting them? It is hard to dissect this information just from documentation. 

Thanks so much for your help,
Angela

**Comment by mattfidler** [comment, marked as answer]
Hi @afee36912 

>  have tried setting nBurn and nEm to set the maximum # of iterations in saem fitting, but I do not have a good understanding of these parameters. What do these parameters do?

These are the two phases of optimization that `saem` uses. Monolix briefly talks about this here:

https://monolix.lixoft.com/graphics/convergence-saem/

The first number of iterations is controlled by the `nBurn`.  In this phase the bouncing around of the parameters should stabilize (or the `saem` algorithm really didn't converge).   If this doesn't happen in your problem you should set this higher.

`nEm` is the number of iterations in the second phase, and should be enough that the values stabilize at the end.

If any of these parameters are not stable, then your model has not converged using `saem`.

``` r
library(nlmixr2)
#> Loading required package: nlmixr2data
f <- function() {
  ini({
    ## You may label each parameter with a comment
    tka <- 0.45 # Log Ka
    tcl <- log(c(0, 2.7, 100)) # Log Cl
    ## This works with interactive models
    ## You may also label the preceding line with label("label text")
    tv <- 3.45; label("log V")
    ## the label("Label name") works with all models
    eta.ka ~ 0.6
    eta.cl ~ 0.3
    eta.v ~ 0.1
    add.sd <- 0.7
  })
  model({
    ka <- exp(tka + eta.ka)
    cl <- exp(tcl + eta.cl)
    v <- exp(tv + eta.v)
    linCmt() ~ add(add.sd)
  })
}

fit <- nlmixr2(f, theo_sd, "saem", control=saemControl(nBurn=500, nEm=750, print=0))
#> ℹ parameter labels from comments will be replaced by 'label()'
#> → loading into symengine environment...
#> → pruning branches (`if`/`else`) of saem model...
#> ✔ done
#> → finding duplicate expressions in saem model...
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> ✔ done
#> using C compiler: ‘gcc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0’
#> ℹ calculate uninformed etas
#> ℹ done
#> rxode2 2.1.3.9000 using 8 threads (see ?getRxThreads)
#>   no cache: create with `rxCreateCache()`
#> Calculating covariance matrix
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> → loading into symengine environment...
#> → pruning branches (`if`/`else`) of saem model...
#> ✔ done
#> → finding duplicate expressions in saem predOnly model 0...
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> → finding duplicate expressions in saem predOnly model 1...
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> → finding duplicate expressions in saem predOnly model 2...
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> ✔ done
#> using C compiler: ‘gcc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0’
#> → Calculating residuals/tables
#> ✔ done
#> → compress origData in nlmixr2 object, save 5952
#> → compress phiM in nlmixr2 object, save 169568
#> → compress parHistData in nlmixr2 object, save 33784
#> → compress saem0 in nlmixr2 object, save 19048

traceplot(fit)
```

![](https://i.imgur.com/AhwJ9DS.png)

<sup>Created on 2024-07-25 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

**Reply by mattfidler** [reply]
> What method do you recommend to optimize parameter fitting? 

`saem` is often quite robust; `focei` is more deterministic, and will try to find the local minimum.

> What should eta values be set to in the guessing stages of nlmixr modeling? 

Typically `saem` likes many etas and large initial values to the `omega` matrices (specified by `eta.cl ~ 1` or something similar.

Conversly, `focei` likes few etas and small initial values.

**Reply by mattfidler** [reply]
You can increase the robustness, memory usage and computation time of `saem` with `nmc` higher than the default of  `3`.  You could also increase the number of nu with a similar effect `nu=c(3, 3, 3)` for instance.

**Reply by afee36912** [reply]
Thanks so much! This is very helpful!
